In [3]:
import numpy as np
import pandas as pd

In [4]:
url = 'https://raw.githubusercontent.com/farisknight13/Machine_Learning/main/drug200.csv'
df = pd.read_csv(url)
df

Age Sex      BP Cholesterol  Na_to_K   Drug
0     23   F    HIGH        HIGH   25.355  DrugY
1     47   M     LOW        HIGH   13.093  drugC
2     47   M     LOW        HIGH   10.114  drugC
3     28   F  NORMAL        HIGH    7.798  drugX
4     61   F     LOW        HIGH   18.043  DrugY
..   ...  ..     ...         ...      ...    ...
195   56   F     LOW        HIGH   11.567  drugC
196   16   M     LOW        HIGH   12.006  drugC
197   52   M  NORMAL        HIGH    9.894  drugX
198   23   M  NORMAL      NORMAL   14.020  drugX
199   40   F     LOW      NORMAL   11.349  drugX

[200 rows x 6 columns]

## **Explain Dataset**
Dataset ที่ใช้เกี่ยวกับการแยกประเภทยาของผู้ป่วย โดยจะแบ่งออกเป็น 5 ประเภท คือ DrugY, drugX, drugA, drugC, drugB <br>
โดยจะแยกจาก อายุของผู้ป่วย, เพศของผู้ป่วย, ระดับความดันโลหิต(ความดันโลหิตแบ่งเป็น ความดันต่ำ, ความดันปกติ, และความดันสูง), <br>ระดับคอเลสเตอรอล(ระดับคอเลสเตอรอลแบ่งเป็น คอเลสเตอรอลปกติ และคอเลสเตอรอลสูง), อัตราส่วนโซเดียมต่อโพแทสเซียมในเลือด

In [5]:
#Change value to dummy value
df['Sex'] = df['Sex'].replace(['M','F'],['0','1'])
df['BP'] = df['BP'].replace(['LOW','NORMAL','HIGH'],['0','1','2'])
df['Cholesterol'] = df['Cholesterol'].replace(['NORMAL','HIGH'],['0','1',])
df['Drug'] = df['Drug'].replace(['DrugY','drugX','drugA','drugC','drugB'],['1','2','3','4','5'])
df

Age Sex BP Cholesterol  Na_to_K Drug
0     23   1  2           1   25.355    1
1     47   0  0           1   13.093    4
2     47   0  0           1   10.114    4
3     28   1  1           1    7.798    2
4     61   1  0           1   18.043    1
..   ...  .. ..         ...      ...  ...
195   56   1  0           1   11.567    4
196   16   0  0           1   12.006    4
197   52   0  1           1    9.894    2
198   23   0  1           0   14.020    2
199   40   1  0           0   11.349    2

[200 rows x 6 columns]

## **Change value to dummy value**
Colume Sex(เพศของผู้ป่วย) แทนค่า M เป็น 0 และ F เป็น 1 <br>
Colume BP(ระดับความดันโลหิต) แทนค่า LOW เป็น 0 , NORMAL เป็น 1 และ HIGH เป็น 2 <br>Colume Cholesterol(ระดับคอเลสเตอรอล) แทนค่า NORMAL เป็น 0 และ HIGH เป็น 1 <br>Colume Drug(ประเภทของยา) แทนค่า DrugY เป็น 1 , drugX เป็น 2 , drugA เป็น 3 , drugC เป็น 4 และ drugB เป็น 5

In [7]:
#split data and drug
X = df.iloc[:,:-1]
y = df.iloc[:,-1]
X

Age Sex BP Cholesterol  Na_to_K
0     23   1  2           1   25.355
1     47   0  0           1   13.093
2     47   0  0           1   10.114
3     28   1  1           1    7.798
4     61   1  0           1   18.043
..   ...  .. ..         ...      ...
195   56   1  0           1   11.567
196   16   0  0           1   12.006
197   52   0  1           1    9.894
198   23   0  1           0   14.020
199   40   1  0           0   11.349

[200 rows x 5 columns]

In [8]:
from sklearn.model_selection import train_test_split
#split Train and Test Data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

## **Direct classification**
classification ที่เลือกใช้คือ K-nearest-neighbors Classifier
โดยทำลองใช้งานโดยเริ่มจาก n=3,5,10 เพิ่มขึ้นตามลำดับ
และดูผลลัพธ์

In [9]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.metrics import accuracy_score

#N=3
classifier_3 = KNeighborsClassifier(n_neighbors=3)
classifier_3.fit(X_train, y_train)
y_pred_3 = classifier_3.predict(X_test)
print("Accuracy N=3 => "+ str(accuracy_score(y_test,y_pred_3)))
#print(confusion_matrix(y_test, y_pred_3))
print(classification_report(y_test, y_pred_3))

#N=5
classifier_5 = KNeighborsClassifier(n_neighbors=5)
classifier_5.fit(X_train, y_train)
y_pred_5 = classifier_5.predict(X_test)
print("Accuracy N=5 => "+ str(accuracy_score(y_test,y_pred_5)))
#print(confusion_matrix(y_test, y_pred_5))
print(classification_report(y_test, y_pred_5))

#N=10
classifier_10 = KNeighborsClassifier(n_neighbors=10)
classifier_10.fit(X_train, y_train)
y_pred_10 = classifier_10.predict(X_test)
print("Accuracy N=10 => "+ str(accuracy_score(y_test,y_pred_10)))
#print(confusion_matrix(y_test, y_pred_10))
print(classification_report(y_test, y_pred_10))

Accuracy N=3 => 0.7333333333333333
              precision    recall  f1-score   support

           1       0.93      0.93      0.93        30
           2       0.61      0.61      0.61        18
           3       0.44      0.80      0.57         5
           4       0.00      0.00      0.00         4
           5       0.33      0.33      0.33         3

    accuracy                           0.73        60
   macro avg       0.46      0.54      0.49        60
weighted avg       0.70      0.73      0.71        60

Accuracy N=5 => 0.7666666666666667
              precision    recall  f1-score   support

           1       0.94      0.97      0.95        30
           2       0.63      0.67      0.65        18
           3       0.50      0.80      0.62         5
           4       0.00      0.00      0.00         4
           5       0.50      0.33      0.40         3

    accuracy                           0.77        60
   macro avg       0.51      0.55      0.52        60
weighte

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## สรุป Direst Classification
หลังจากนำ dataset ไปทำ Direct Classification พบว่ายิ่งใช้จำนวน N เพิ่มมากขึ้นมาโอกาศทำให้การทำนายผิดพลาดได้ เพราะจำนวนข้อมูลที่นำเข้ามาแต่ละ class มีไม่เท่ากันจึงอาจไปเจอ class ที่มีตัวนวนตัวอยู่ใกล้มากกว่าและทำให้การทำนายผิดพลายได้ และการที่จำนวน N น้อยเกินไปก็อาจทำให้ไม่ความแม่นยำลดลงได้เช่นเดียวกัน เพราะตัวที่อยู่ใกล้อาจมีจำนวนข้อมูลของแต่ละ class เท่ากัน จึงต้องเลือกใช้จำนวน N ที่เหมาะสมกับแต่ละ dataset โดยใน Dataset นี้ คือ N=5 จะให้ Accuracy เท่ากับ 0.76666667

## **Traditional Classification**
Model ที่ใช้ในการทดลองมี Decision Tree และ Random Forest 

In [10]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

#DecisionTree
clf_NDT = DecisionTreeClassifier()
clf_NDT = clf_NDT.fit(X_train,y_train)
y_pred_NDT = clf_NDT.predict(X_test)
print('Accuracy DecisionTree =',accuracy_score(y_test,y_pred_NDT))
print(confusion_matrix(y_pred_NDT, y_test))

#RandomForest
clf_NRF = RandomForestClassifier()
clf_NRF = clf_NRF.fit(X_train,y_train)
y_pred_NRF = clf_NRF.predict(X_test)
print('Accuracy RandomForest =',accuracy_score(y_test,y_pred_NRF))
print(confusion_matrix(y_test, y_pred_NRF))

Accuracy DecisionTree = 1.0
[[30  0  0  0  0]
 [ 0 18  0  0  0]
 [ 0  0  5  0  0]
 [ 0  0  0  4  0]
 [ 0  0  0  0  3]]
Accuracy RandomForest = 1.0
[[30  0  0  0  0]
 [ 0 18  0  0  0]
 [ 0  0  5  0  0]
 [ 0  0  0  4  0]
 [ 0  0  0  0  3]]


## สรุป Traditional Classification
ผลการนำ dataset ไปเข้า model Decision Tree และ Random Forest ได้ค่า Accuracy และ confusion matrix เท่ากัน ค่า Accuracy เท่ากับ 1 ซึ่งหมายความว่า model มีการทำนายที่แม่นยำ ไม่มีข้อผิดพลาดเลย ซึ่ง Decision Tree และ Random Forest เป็นการนำ dataset ไปเข้า model เพื่อสร้างต้นไม้ออกมา เพื่อแยกประเภทของแต่ละclass ซึ่งแสดงว่า dataset นี้เหมาะกับการทำ Traditional Classification แบบ Decision Tree และ Random Forest

## **Deep Learning**
เลือกใช้ model แบบ Multi-layer Perceptron ซึ่งเป็น Model Neural network ประเภทหนึ่ง และได้ทำการทดลองด้วยการปรับค่า hidden layer เท่ากับ 5, 6, 8, 9 ตามลำดับและดูผลลัพธ์

In [34]:
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import cross_val_score

#hidden layer = 5
clf_MLP_5 = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(5), random_state=1)
clf_MLP_5.fit(X_train, y_train)
y_pred_MLP_5 = clf_MLP_5.predict(X_test)
print('Accuracy MLPClassifier_5 =',accuracy_score(y_test,y_pred_MLP_5))
print(confusion_matrix(y_test, y_pred_MLP_5))

#hidden layer = 6
clf_MLP_6 = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(6), random_state=1)
clf_MLP_6.fit(X_train, y_train)
y_pred_MLP_6 = clf_MLP_6.predict(X_test)
print('Accuracy MLPClassifier_6 =',accuracy_score(y_test,y_pred_MLP_6))
print(confusion_matrix(y_test, y_pred_MLP_6))

#hidden layer = 8
clf_MLP_8 = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(8), random_state=1)
clf_MLP_8.fit(X_train, y_train)
y_pred_MLP_8 = clf_MLP_8.predict(X_test)
print('Accuracy MLPClassifier_8 =',accuracy_score(y_test,y_pred_MLP_8))
print(confusion_matrix(y_test, y_pred_MLP_8))

#hidden layer = 8
clf_MLP_9 = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(9), random_state=1)
clf_MLP_9.fit(X_train, y_train)
y_pred_MLP_9 = clf_MLP_9.predict(X_test)
print('Accuracy MLPClassifier_9 =',accuracy_score(y_test,y_pred_MLP_9))
print(confusion_matrix(y_test, y_pred_MLP_9))

Accuracy MLPClassifier_5 = 0.5
[[30  0  0  0  0]
 [18  0  0  0  0]
 [ 5  0  0  0  0]
 [ 4  0  0  0  0]
 [ 3  0  0  0  0]]
Accuracy MLPClassifier_6 = 0.75
[[29  1  0  0  0]
 [ 0 16  2  0  0]
 [ 0  5  0  0  0]
 [ 1  3  0  0  0]
 [ 0  3  0  0  0]]
Accuracy MLPClassifier_8 = 0.9333333333333333
[[29  1  0  0  0]
 [ 1 15  0  1  1]
 [ 0  0  5  0  0]
 [ 0  0  0  4  0]
 [ 0  0  0  0  3]]
Accuracy MLPClassifier_9 = 0.6166666666666667
[[27  3  0  0  0]
 [ 3  6  9  0  0]
 [ 0  1  4  0  0]
 [ 1  2  1  0  0]
 [ 0  1  2  0  0]]


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:470: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


## สรุป Deep Learning
Deep Learning ได้ Accuracy มากที่สุดเท่ากับ 0.9333 ในการใส่พารามิเตอร์ hidden layer เท่ากับ 8 โดยจากการทดลองทำให้เห็นว่า dataset นี้เหมาะกับการใช้ hidden layer เท่ากับ 8 เพราะเมื่อเพิ่มจำนวน hidden layer ไปเป็น 9 ทำให้ค่า Accuracy ลดลง